packages/browser/src/profiling/hubextensions.ts

```diff
@@ -0,0 +1,257 @@
+import { getCurrentHub, getMainCarrier } from '@sentry/core';
+import type { CustomSamplingContext, Hub, Transaction, TransactionContext } from '@sentry/types';
+import { logger, uuid4 } from '@sentry/utils';
+
+import { WINDOW } from '../helpers';
+import type { JSSelfProfile, JSSelfProfiler, ProcessedJSSelfProfile } from './jsSelfProfiling';
+import { sendProfile } from './sendProfile';
+
+// Max profile duration.
+const MAX_PROFILE_DURATION_MS = 30_000;
+
+// While we experiment, per transaction sampling interval will be more flexible to work with.
+type StartTransaction = (
+  this: Hub,
+  transactionContext: TransactionContext,
+  customSamplingContext?: CustomSamplingContext,
+) => Transaction | undefined;
+
+/**
+ * Check if profiler constructor is available.
+ * @param maybeProfiler
+ */
+function isJSProfilerSupported(maybeProfiler: unknown): maybeProfiler is typeof JSSelfProfiler {
+  return typeof maybeProfiler === 'function';
+}
+
+/**
+ * Safety wrapper for startTransaction for the unlikely case that transaction starts before tracing is imported -
+ * if that happens we want to avoid throwing an error from profiling code.
+ * see https://github.com/getsentry/sentry-javascript/issues/4731.
+ */
+export function onProfilingStartRouteTransaction(transaction: Transaction | undefined): Transaction | undefined {
```

Comment:
m: Since this is the function that is actually part of the public API I suggest we make this name more clear. I would name this function `wrapTransactionWithProfiling` and the one that is currently named `wrapTransactionWithProfiling` something like `_wrapTransactionWithProfiling`. Wdyt?

Reply:
I think it's actually clearer like this - the reason for that is that we only want to export this fn for the experimental browser tracing option and not have someone wrap any other transaction with profiling. To me, wrapTransactionWithProfiling implies "any" transaction which is something that we want to discourage